compiler/rustc_errors/src/lib.rs

```diff
@@ -882,16 +884,15 @@ impl DiagCtxt {
         if treat_next_err_as_bug {
             self.span_bug(sp, msg);
         }
-        DiagnosticBuilder::<ErrorGuaranteed>::new(self, DelayedBug, msg).with_span(sp).emit()
+        DiagnosticBuilder::<ErrorGuaranteed>::new(self, DelayedBug(DelayedBugKind::Normal), msg)
+            .with_span(sp)
+            .emit()
     }
 
     // FIXME(eddyb) note the comment inside `impl Drop for DiagCtxtInner`, that's
     // where the explanation of what "good path" is (also, it should be renamed).
     pub fn good_path_delayed_bug(&self, msg: impl Into<DiagnosticMessage>) {
-        let mut inner = self.inner.borrow_mut();
-        let diagnostic = Diagnostic::new(DelayedBug, msg);
-        let backtrace = std::backtrace::Backtrace::capture();
-        inner.good_path_delayed_bugs.push(DelayedDiagnostic::with_backtrace(diagnostic, backtrace));
+        DiagnosticBuilder::<()>::new(self, DelayedBug(DelayedBugKind::GoodPath), msg).emit()
```

Comment:
Should be `ErrorGuaranteed` instead of `()`, and also this method can return `ErrorGuaranteed` just like `span_delayed_bug`.

Reply:
It is not correct to return `ErrorGuaranteed` from `good_path_bug`.

If I delayed a good path bug, emitted a warning-level diagnostic to suppress the good path bugs, then constructed a `Ty::new_error` from that `ErrorGuranteed`, I could end up with a `TyKind::Error` in codegen. That's not sound.

Reply:
Hmm, ok, I see what you're saying. I hate good path delayed bugs! I'd love to get rid of them, I think the one in `TypeErrCtxt::drop` is bogus and should be something else (though I haven't worked out what, yet) and the other one in `trimmed_def_paths` I wonder if it's even worth having.

Reply:
Ok, but for now, I'd rather keep what I have so we don't accidentally expose new ways for people to make `ErrorGuaranteed`s out of nothing. Creating an `ErrorGuaranteed` when you shouldn't have one could lead to really bad bugs that can go unnoticed and it's not worth a minor cleanup imo.

It's possible we could construct the diagnostic as a `DiagnosticBuilder<ErrorGuranteed>`, but the `emit` call still should return `None` for clarity, and the `fn good_path_bug` user-facing API still needs to return `()` for **correctness**, so I don't see the harm in just making it act function as a `DiagnosticBuilder<()>` always.